### PR TITLE
Introduced AndFinally behavior to Try

### PR DESCRIPTION
**Pull request by vicmosin**
Motivation is to provide java try's finally behavior independent from the result of the `Try`.

### CODE REVIEW

File: javaslang/src/test/java/javaslang/control/TryTest.java [hunk 1]
```diff
@@ -1044,6 +1051,50 @@ public void shouldReturnRecoveredValueOnFailure(){
...
+        });
+
+        assertThat(count.getValue()).isEqualTo(1);
+    }
+
+    @Test
+    public void shouldExecuteAndFinallyTryOnSuccess(){
+        MutableInteger count = new MutableInteger(0);
+
+        Try.success(count).andFinallyTry(() -> {
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/control/TryTest.java]: \[FindBugs\] INFO: RV: Return value of Try\.andFinallyTry\(Runnable\) ignored, is this OK in javaslang\.control\.TryTest\.shouldExecuteAndFinallyTryOnSuccess\(\)

File: javaslang/src/test/java/javaslang/control/TryTest.java [hunk 2]
```diff
@@ -1044,6 +1051,50 @@ public void shouldReturnRecoveredValueOnFailure(){
...
+        });
+
+        assertThat(count.getValue()).isEqualTo(1);
+    }
+
+    @Test
+    public void shouldExecuteAndFinallyTryOnFailure(){
+        MutableInteger count = new MutableInteger(0);
+
+        Try.run(() -> {throw new IllegalStateException(FAILURE);}).andFinallyTry(() -> {
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/control/TryTest.java]: \[FindBugs\] INFO: RV: Return value of Try\.andFinallyTry\(Runnable\) ignored, is this OK in javaslang\.control\.TryTest\.shouldExecuteAndFinallyTryOnFailure\(\)

File: javaslang/src/test/java/javaslang/control/TryTest.java [hunk 3]
```diff
@@ -1044,6 +1051,44 @@ public void shouldReturnRecoveredValueOnFailure(){
...
+        Try.run(() -> count.setValue(0)).andFinally(() -> count.setValue(1));
+
+        assertThat(count.getValue()).isEqualTo(1);
+    }
+
+    @Test
+    public void shouldExecuteAndFinallyTryOnSuccess(){
+        MutableInteger count = new MutableInteger(0);
+
+        Try.run(() -> count.setValue(0)).andFinallyTry(() -> count.setValue(1));
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/control/TryTest.java]: \[FindBugs\] INFO: RV: Return value of Try\.andFinallyTry\(Runnable\) ignored, is this OK in javaslang\.control\.TryTest\.shouldExecuteAndFinallyTryOnSuccess\(\)

File: javaslang/src/test/java/javaslang/control/TryTest.java [hunk 4]
```diff
@@ -1044,6 +1051,44 @@ public void shouldReturnRecoveredValueOnFailure(){
...
+
+        assertThat(count.getValue()).isEqualTo(1);
+    }
+
+    @Test
+    public void shouldExecuteAndFinallyTryOnFailure(){
+        MutableInteger count = new MutableInteger(0);
+
+        Try.run(() -> {throw new IllegalStateException(FAILURE);})
+            .andFinallyTry(() -> count.setValue(1));
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/control/TryTest.java]: \[FindBugs\] INFO: RV: Return value of Try\.andFinallyTry\(Runnable\) ignored, is this OK in javaslang\.control\.TryTest\.shouldExecuteAndFinallyTryOnFailure\(\)

File: javaslang/src/test/java/javaslang/control/TryTest.java (hunk 3 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/control/TryTest.java]: \[FindBugs\] INFO: RV: Return value of Try\.andFinallyTry\(Try$CheckedRunnable\) ignored, is this OK in javaslang\.control\.TryTest\.shouldExecuteAndFinallyTryOnSuccess\(\)

File: javaslang/src/test/java/javaslang/control/TryTest.java (hunk 4 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/control/TryTest.java]: \[FindBugs\] INFO: RV: Return value of Try\.andFinallyTry\(Try$CheckedRunnable\) ignored, is this OK in javaslang\.control\.TryTest\.shouldExecuteAndFinallyTryOnFailure\(\)

**Comment by danieldietrich** [comment]
Great addition, kose and tests look good 👍🏼

File: javaslang/src/test/java/javaslang/control/TryTest.java (hunk 3 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/control/TryTest.java]: \[FindBugs\] INFO: RV: Return value of Try\.andFinallyTry\(Try$CheckedRunnable\) ignored, is this OK in javaslang\.control\.TryTest\.shouldExecuteAndFinallyTryOnSuccess\(\)

File: javaslang/src/test/java/javaslang/control/TryTest.java (hunk 4 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/control/TryTest.java]: \[FindBugs\] INFO: RV: Return value of Try\.andFinallyTry\(Try$CheckedRunnable\) ignored, is this OK in javaslang\.control\.TryTest\.shouldExecuteAndFinallyTryOnFailure\(\)

File: javaslang/src/main/java/javaslang/control/Try.java [hunk 5]
```diff
@@ -719,6 +719,35 @@ default void orElseRun(Consumer<? super Throwable> action) {
...
+     * @return this {@code Try}.
+     * @throws NullPointerException if {@code runnable} is null
+     */
+    default Try<T> andFinallyTry(CheckedRunnable runnable) {
+        Objects.requireNonNull(runnable, "runnable is null");
+        try {
+            runnable.run();
+            return this;
+        } catch (Throwable t) {
+            return new Failure<>(t);
```
Comment (ruslansennov) [inline javaslang/src/main/java/javaslang/control/Try.java]: not tested?

**Comment by danieldietrich** [comment]
Thank you!

**Comment by vicmosin** [comment]
Any time ;-)

File: javaslang/src/main/java/javaslang/control/Try.java (hunk 5 above)
Comment (danieldietrich) [inline javaslang/src/main/java/javaslang/control/Try.java]: @vicmosin can you provide a test please? I've overseen it - I'm on a mobile phone...

**Pull request merged**